### PR TITLE
add checks for NSR to overlap a split read

### PR DESCRIPTION
**Pull request by nickhsmith**
Make sure for a valid FDS object that the non-spliced reads overlap with at least 1 spliced read donor or acceptor.

Enforce this during the external splice count merging

### CODE REVIEW

**Comment by nickhsmith** [comment]
I don't understand the devel release failures

File: .github/helperScripts/setupEnv.R [hunk 1]
```diff
@@ -17,7 +17,7 @@ installIfReq <- function(p, Ncpus=NCPUS, ...){
 # install Bioconductor
 if(!requireNamespace("BiocManager", quietly=TRUE))
     install.packages("BiocManager", Ncpus=NCPUS)
-BiocManager::install("BiocVersion", version=BIOC_VERSION)
+BiocManager::install("BiocVersion", version=BIOC_VERSION,ask = FALSE)
```
Comment (nickhsmith) [inline .github/helperScripts/setupEnv.R:20]: important for the `devel` BIOC version

File: .github/workflows/r.yml [hunk 2]
```diff
@@ -16,9 +16,9 @@ jobs:
       fail-fast: false
       matrix:
         config:
-          - {os: windows-latest, r: 'release', bioc: '3.14'}
-          - {os: macOS-latest,   r: 'release', bioc: '3.14'} # Due to gfortran installation issues
-          - {os: ubuntu-20.04,   r: 'release', bioc: '3.14',  rspm: "https://packagemanager.rstudio.com/cran/__linux__/focal/latest"}
+          - {os: windows-latest, r: '4.2', bioc: '3.15'}
+          - {os: macOS-latest,   r: '4.2', bioc: '3.15'} # Due to gfortran installation issues
```
Comment (c-mertes) [inline .github/workflows/r.yml]:
```suggestion
          - {os: macOS-latest,   r: '4.2', bioc: '3.15'}
```

File: R/FraserDataSet-class.R [hunk 3]
```diff
@@ -134,6 +134,25 @@ validateAssays <- function(object){
...
 }
 
+# for non-empty fds objects check if non-spliced reads are overlapping with at least 1 donor/acceptor site
+validateNonSplicedReadsSanity <- function(object){
+    # fds object must have samples and junctions
+    if(all(dim(object) > c(0,0))){
+
+        # fds object must be annotated with start/end/spliceSite indexes
+        if(any("startID" == names(rowData(object))) && any("endID" == names(rowData(object))) &&
+           any("spliceSiteID" == names(rowData(object@nonSplicedReads))) ){
```
Comment (c-mertes) [inline R/FraserDataSet-class.R]:
```suggestion
        if("startID" %in% names(mcols(object)) && "endID" %in% names(mcols(object)) &&
           "spliceSiteID" %in% names(mcols(object, type="theta"))){
```

File: R/FraserDataSet-class.R [hunk 4]
```diff
@@ -134,6 +134,25 @@ validateAssays <- function(object){
...
+
+                # check that every spliceSiteID matches either a start or end index
+                if(length(intersect(rowData(object@nonSplicedReads)$spliceSiteID, c(rowData(object)$startID,rowData(object)$endID)))
+                      != dim(object@nonSplicedReads)[1]){
```
Comment (c-mertes) [inline R/FraserDataSet-class.R]:
```suggestion
                if(length(intersect(mcols(object, type="theta")$spliceSiteID,  unlist(mcols(object)[, c("startID", "endID")]))
                      != nrow(nonSplicedReads(object))){
```

File: R/mergeExternalData.R [hunk 5]
```diff
@@ -118,48 +118,60 @@ mergeExternalData <- function(fds, countFiles, sampleIDs, annotation=NULL){
...
+             elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
+             metadata = metadata(fds)
+  )
+  
+  # 
+  # compute new psi values
+  # 
+  ans <- calculatePSIValues(ans)
+  
+  ans
```
Comment (c-mertes) [inline R/mergeExternalData.R]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

Comment (c-mertes) [inline R/mergeExternalData.R:176]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

Comment (c-mertes) [inline R/mergeExternalData.R]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

Comment (c-mertes) [inline R/mergeExternalData.R:176]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

Comment (c-mertes) [inline R/mergeExternalData.R:176]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

Comment (c-mertes) [inline R/mergeExternalData.R:176]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

File: R/mergeExternalData.R [hunk 6]
```diff
@@ -118,48 +118,60 @@ mergeExternalData <- function(fds, countFiles, sampleIDs, annotation=NULL){
...
+  NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
+  
+  # Find the overlaps of the NSR with SR after merging/filtering
+  NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
+  
+  # Only take NSR that have at least 1 split read over the same junction.
+  NSR_ranges <- NSR_ranges[NSR_index]
+  newCtsK_theta <- newCtsK_theta[NSR_index,]
+  newCtsN_theta <- newCtsN_theta[NSR_index,]
+
```
Comment (c-mertes) [inline R/mergeExternalData.R:143]:
```suggestion
    SR_ranges <- rowRanges(fds)[from(ov),c("startID", "endID")]
    
    
    #
    # merge theta data
    #
    # find overlap
    ovss <- findOverlaps(rowRanges(fds, type="theta"),
            extCts[['k_theta']], type="equal")
    
    newCtsK_theta <- extractExtData(fds, K, "theta", ovss, extCts, "k_theta")
    newCtsN_theta <- extractExtData(fds, N, "theta", ovss, extCts, "n_theta")
    NSR_ranges <- rowRanges(fds, type="theta")[from(ovss),c("spliceSiteID", "type")]
    
    # Find the overlaps of the NSR with SR after merging/filtering
    NSR_index <- which(NSR_ranges$spliceSiteID %in% c(SR_ranges$startID, SR_ranges$endID))
    
    # Only take NSR that have at least 1 split read over the same junction.
    NSR_ranges <- NSR_ranges[NSR_index]
    newCtsK_theta <- newCtsK_theta[NSR_index,]
    newCtsN_theta <- newCtsN_theta[NSR_index,]
    
```

File: R/mergeExternalData.R (hunk 5 above)
Comment (c-mertes) [inline R/mergeExternalData.R:176]:
```suggestion
    #
    # finalize merged FraserDataObject
    #
    nsr <- SummarizedExperiment(
            colData = newColData,
            assays = SimpleList(
                    rawCountsSS = newCtsK_theta,
                    rawOtherCounts_theta = (newCtsN_theta - newCtsK_theta)),
            rowRanges= NSR_ranges
    )
    
    ans <- new("FraserDataSet", 
            name = name(fds),
            bamParam = scanBamParam(fds),
            strandSpecific = strandSpecific(fds),
            workingDir = workingDir(fds),
            colData = newColData,
            assays = Assays(SimpleList(
                    rawCountsJ = newCtsK_J,
                    rawOtherCounts_psi5 = newCtsN_psi5 - newCtsK_J,
                    rawOtherCounts_psi3 = newCtsN_psi3 - newCtsK_J)),
            nonSplicedReads = nsr,
            rowRanges = SR_ranges,
            elementMetadata = DataFrame(newCtsK_J[,integer(0)]),
            metadata = metadata(fds)
    )
    
    # 
    # compute new psi values
    # 
    ans <- calculatePSIValues(ans)
    
    ans
```

**Review by c-mertes — APPROVED** [review]
thanks for the push @nickhsmith this looks good. I will merge it and push it to a new version.

**Pull request merged**